source/extensions/filters/network/mysql_proxy/mysql_codec_clogin_resp.h

```diff
@@ -1,38 +1,127 @@
 #pragma once
-#include <cstdint>
+
+#include "envoy/buffer/buffer.h"
 
 #include "common/buffer/buffer_impl.h"
 
 #include "extensions/filters/network/mysql_proxy/mysql_codec.h"
+#include "extensions/filters/network/mysql_proxy/mysql_codec_clogin.h"
 
 namespace Envoy {
 namespace Extensions {
 namespace NetworkFilters {
 namespace MySQLProxy {
 
+class ClientLoginResponse;
+class AuthMoreMessage;
+class OkMessage;
+class ErrMessage;
+class AuthSwitchMessage;
+using ClientLoginResponsePtr = std::unique_ptr<ClientLoginResponse>;
+
+// ClientLoginResponse could be
+// Protocol::OldAuthSwitchRequest, Protocol::AuthSwitchRequest when server wants switch auth
+// method or OK_Packet, ERR_Packet when server auth ok or error
 class ClientLoginResponse : public MySQLCodec {
 public:
+  ClientLoginResponse() = default;
+  ClientLoginResponse(const ClientLoginResponse& other);
   // MySQLCodec
-  int parseMessage(Buffer::Instance& buffer, uint32_t len) override;
-  std::string encode() override;
+  DecodeStatus parseMessage(Buffer::Instance& buffer, uint32_t len) override;
+  void encode(Buffer::Instance&) override;
 
   uint8_t getRespCode() const { return resp_code_; }
-  uint8_t getAffectedRows() const { return affected_rows_; }
-  uint8_t getLastInsertId() const { return last_insert_id_; }
-  uint16_t getServerStatus() const { return server_status_; }
+  void setRespCode(uint8_t resp_code) { resp_code_ = resp_code; }
+  void initMessage(uint8_t resp_code);
+
+  AuthMoreMessage& asAuthMoreMessage();
+  AuthSwitchMessage& asAuthSwitchMessage();
+  OkMessage& asOkMessage();
+  ErrMessage& asErrMessage();
+
+private:
+  uint8_t resp_code_{MYSQL_RESP_UNKNOWN};
+  ClientLoginResponsePtr message_;
```

Comment:
While it works, but notice that `message_` content, which is derived from `ClientLoginResponse`, also has `resp_code_` which is not used and initialized  to MYSQL_RESP_UNKNOWN. I think that you should restructure a bit. Deriving AuthMoreMessage, ErrMessage, OKMessage from ClientLoginResponse simplified things. I think you should remove `message_` from here and leave only _resp_code__. Then in `onClientLoginResponse` callback peek or read 1 byte and allocate AuthMoreMessage, AuthSwitchMessage, OKMessage and ErrMessage. Something like:


```
     resp_code = buf.peekLE<uint8_t>(...)

     switch (resp_code) {

     case MYSQL_RESP_OK:

     message_ = std::make_unique<OkMessage>();

     .....
```
Initialize `resp_code_` in a constructor of each message.
After that appropriate parse method will be called.

When you do it, `as...` methods: `asAuthMoreMesage`, `asAuthSwitchMessage`, etc should disappear.

Reply:
Thks! Fixed as you say.